src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp

```diff
@@ -145,24 +269,119 @@ bool ShenandoahAdaptiveHeuristics::should_start_gc() const {
+                       byte_size_in_proper_unit(penalties),           proper_unit_for_byte_size(penalties),
+                       byte_size_in_proper_unit(allocation_headroom), proper_unit_for_byte_size(allocation_headroom));
+
+    _last_trigger = RATE;
     return true;
   }
 
-  return ShenandoahHeuristics::should_start_gc();
+  double instantaneous_rate = _allocation_rate.instantaneous_rate(bytes_allocated_since_gc_start);
+  if (_allocation_rate.is_spiking(instantaneous_rate) && average_cycle_seconds > allocation_headroom / instantaneous_rate) {
+    log_info(gc)("Trigger: Instantaneous allocation rate (%.0f %sB/s) will deplete free headroom (" SIZE_FORMAT "%s) before average time (%.2f ms) to complete GC cycle.",
+                 byte_size_in_proper_unit(instantaneous_rate), proper_unit_for_byte_size(instantaneous_rate),
+                 byte_size_in_proper_unit(allocation_headroom), proper_unit_for_byte_size(allocation_headroom),
+                 average_cycle_seconds * 1000);
+    _last_trigger = SPIKE;
+    return true;
+  }
+
+  return false;
+}
+
+void ShenandoahAdaptiveHeuristics::adjust_last_trigger_parameters(double amount) {
+  switch (_last_trigger) {
+    case RATE:
+      adjust_margin_of_error(amount);
+      break;
+    case SPIKE:
+      adjust_spike_threshold(amount);
+      break;
+    case OTHER:
+      // nothing to adjust here.
+      break;
+    default:
+      ShouldNotReachHere();
+  }
+}
+
+void ShenandoahAdaptiveHeuristics::adjust_margin_of_error(double amount) {
+  _margin_of_error_sd = saturate(_margin_of_error_sd + amount, MINIMUM_CONFIDENCE, MAXIMUM_CONFIDENCE);
+  log_debug(gc, ergo)("Margin of error now %.2f", _margin_of_error_sd);
+}
+
+void ShenandoahAdaptiveHeuristics::adjust_spike_threshold(double amount) {
+  _spike_threshold_sd = saturate(_spike_threshold_sd - amount, MINIMUM_CONFIDENCE, MAXIMUM_CONFIDENCE);
+  log_debug(gc, ergo)("Spike threshold now: %.2f", _spike_threshold_sd);
+}
+
+ShenandoahAllocationRate::ShenandoahAllocationRate(ShenandoahAdaptiveHeuristics *heuristics) :
+  _heuristics(heuristics),
+  _last_sample_time(os::javaTimeNanos()),
+  _last_sample_value(0),
+  _interval_ns(NANOUNITS / ShenandoahAdaptiveSampleFrequencyHz),
+  _rate(ShenandoahAdaptiveSampleSizeSeconds * ShenandoahAdaptiveSampleFrequencyHz, ShenandoahAdaptiveDecayFactor),
+  _rate_avg(ShenandoahAdaptiveSampleSizeSeconds * ShenandoahAdaptiveSampleFrequencyHz, ShenandoahAdaptiveDecayFactor) {
+}
+
+void ShenandoahAllocationRate::sample(size_t bytes_allocated_since_gc_start) {
+  jlong now = os::javaTimeNanos();
+  if (now - _last_sample_time > _interval_ns) {
+    if (bytes_allocated_since_gc_start > _last_sample_value) {
+      size_t allocation_delta = bytes_allocated_since_gc_start - _last_sample_value;
+      size_t time_delta_ns = now - _last_sample_time;
+      double alloc_bytes_per_second = ((double) allocation_delta * NANOUNITS) / time_delta_ns;
```

Comment:
I wonder if you can just call `instantaneous_rate()` here. Note the comment about that method too.